tests/fixtures/projects/demo-package/requirements.txt

```diff
@@ -1,14 +1,22 @@
 click==7.1.2 \
     --hash=sha256:dacca89f4bfadd5de3d7489b7c8a566eee0d3676333fbb50030263894c38c0dc \
+    --hash=sha256:dacca89f4bfadd5de3d7489b7c8a566eee0d3676333fbb50030263894c38c0dc \
```

Comment:
The hash seems duplicating, maybe some bug?

Reply:
i noticed that yesterday to.
this is actually due to: https://github.com/pdm-project/pdm/blob/main/pdm/project/config.py#L36-L39

since there are two default-repositories, there are two hashes for each file (one for each repository).

the way to prevent duplicate hashes would be using a `set` [here](https://github.com/pdm-project/pdm/blob/main/pdm/formats/requirements.py#L190). but that would change the order of hashes.

Reply:
> this is actually due to: https://github.com/pdm-project/pdm/blob/main/pdm/project/config.py#L36-L39

I doubt that, though a bit confusing, the default repositories are for **uploading packages**, and have nothing to do with downloading and fetching

Reply:
if you take a look at the generated [`pdm.lock`](https://github.com/pdm-project/pdm/blob/aa272a97957356155523148e2cbc138e05322bdb/tests/fixtures/projects/demo-package/pdm.lock#L52-L55) it also contains files from both repositories

Reply:
Should we prefer the link from the prior source(i.e. perform deduplication against filename, the first comes wins)?
The extra sources act as complementary sources. Only those not found in the prior sources take the link from these sources.

Reply:
ah yeah you're right. it's due to `pyproject.toml` in the example projects.

i think we should only do deduplication of hashes in the `requirements.txt`
i see value in having multiple urls for the same package in the `pdm.lock` as it allows fallback. imagine the case where the first source is available at lock-time, not not available at install time.

Reply:
>imagine the case where the first source is available at lock-time, not not available at install time.

Sounds sensible, then we will implement the behavior I mentioned in the gist:

> When installing from the lock file, take the links stored with the lock file.

Reply:
First, of course
Second, of course

Reply:
done